Sources/AsyncHTTPClient/ConnectionPool/ChannelHandler/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,204 @@
+        case .connectSent(let timeout), .headReceived(let timeout):
+            timeout.cancel()
+            let error = HTTPClientError.remoteConnectionClosed
+            self.state = .failed(error)
+            self.proxyEstablishedPromise?.fail(error)
+            context.fireErrorCaught(error)
+        case .failed:
+            break
+        case .completed:
+            break
+        }
+    }
+
+    func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        preconditionFailure("We don't support outgoing traffic during HTTP Proxy update.")
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head(let head):
+            guard case .connectSent(let promise) = self.state else {
+                preconditionFailure("HTTPDecoder should throw an error, if we have not send a request")
+            }
+
+            switch head.status.code {
+            case 200..<300:
+                // Any 2xx (Successful) response indicates that the sender (and all
+                // inbound proxies) will switch to tunnel mode immediately after the
+                // blank line that concludes the successful response's header section
+                self.state = .headReceived(promise)
+            case 407:
+                let error = HTTPClientError.proxyAuthenticationRequired
+                self.state = .failed(error)
+                self.proxyEstablishedPromise?.fail(error)
+                context.fireErrorCaught(error)
+                context.close(mode: .all, promise: nil)
+
+            default:
+                // Any response other than a successful response
+                // indicates that the tunnel has not yet been formed and that the
+                // connection remains governed by HTTP.
+                let error = HTTPClientError.invalidProxyResponse
+                self.state = .failed(error)
+                self.proxyEstablishedPromise?.fail(error)
+                context.fireErrorCaught(error)
+                context.close(mode: .all, promise: nil)
+            }
+        case .body:
+            switch self.state {
+            case .headReceived(let timeout):
+                timeout.cancel()
+                // we don't expect a body
+                let error = HTTPClientError.invalidProxyResponse
+                self.state = .failed(error)
+                self.proxyEstablishedPromise?.fail(error)
+                context.fireErrorCaught(error)
+                context.close(mode: .all, promise: nil)
+
+            case .failed:
+                // ran into an error before... ignore this one
+                break
+            case .completed, .connectSent, .initialized:
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+
+        case .end:
+            switch self.state {
```

Comment:
I still think it would be really beneficial to break this big-boy up a little... Interested to see if @Lukasa agrees

(this isn't a blocker in any way)

Reply:
Yeah, probably a good idea.